fsck/fsck.c

```diff
@@ -775,24 +790,51 @@ static int restore_boot_region(struct exfat_blk_dev *bd)
 
 static int exfat_boot_region_check(struct exfat *exfat, struct pbr **bs)
 {
+	struct pbr *boot_sect;
+	unsigned int sect_size;
 	int ret;
 
-	ret = read_boot_region(exfat->blk_dev, bs, BOOT_SEC_IDX);
+	/* First, find out the exfat sector size */
+	boot_sect = malloc(sizeof(*boot_sect));
+	if (boot_sect == NULL)
+		return -ENOMEM;
+
+	if (exfat_read(exfat->blk_dev->dev_fd, boot_sect,
+		       sizeof(*boot_sect), 0) != (ssize_t)sizeof(*boot_sect)) {
+		exfat_err("failed to read Main boot sector\n");
+		return -EIO;
+	}
+
+	sect_size = 1 << boot_sect->bsx.sect_size_bits;
+	free(boot_sect);
+
+	/* check boot regions */
+	ret = read_boot_region(exfat->blk_dev, bs, true, sect_size, false);
 	if (ret == -EINVAL && exfat_repair_ask(exfat, ER_BS_BOOT_REGION,
 				"boot region is corrupted. try to restore the region from backup"
 				)) {
-		ret = read_boot_region(exfat->blk_dev, bs, BACKUP_BOOT_SEC_IDX);
-		if (ret < 0) {
-			exfat_err("backup boot region is also corrupted\n");
-			return ret;
-		}
-		ret = restore_boot_region(exfat->blk_dev);
-		if (ret < 0) {
-			exfat_err("failed to restore boot region from backup\n");
-			free(*bs);
-			*bs = NULL;
-			return ret;
+		const unsigned int sector_sizes[] = {512, 4096, 1024, 2048};
```

Comment:
we have got exfat sector size from boot sector above. Why do we need to check it using 4 kinds of sector size ?

Reply:
Because we don't trust the sector size of the main boot sector. it seems better to start the loop with the sector size of the main if it is valid.

Reply:
main boot sector should be recover using backup's one? You are saying backup boot sector can even not be trusted ? Have you checked a such case with windows fsck ? I want to know whether windows fsck can recover or not.